core/state_transition.go

```diff
@@ -75,6 +68,24 @@ type Message interface {
 	Data() []byte
 }
 
+// ExecutionResult includes all output after executing given evm message
+// no matter the execution itself is successful or not.
+type ExecutionResult struct {
+	UsedGas      uint64 // Total used gas but include the refunded gas
+	Err          error  // Any error encountered during the exection(listed in core/vm/errors.go)
+	Result       []byte // Returned value of the calling function
+	RevertReason []byte // Reason to perform revert thrown by solidity code
```

Comment:
So, the idea is to return a json-struct that looks like this, I get it. However, the EVM in specification and implementation does not distinguish between `returndata` and `revertdata`. It's all accessed via `RETURNDATA` opcode -- the only difference is that one get's populated via `RETURN` and the other via `REVERT`. So I think treating them as two separate things in `ExecutionResult` does not quite fit.

IMO it would be better to have only `Result []byte`, and when eventually marshalling to json, we can make it look like there was in fact two different things, if that's what is needed to be in agreement with parity. 

Also, nilpick, instead of `... by solidity code`, might be better to say `Data supplied with revert opcode`. The evm isn't solidity-aware.

Reply:
I merge the `Result` and `RevertReason` into `Returndata`. But I offer two more APIs to retrieve `Return` or `Revert` data which is corresponding with `RETURN` and `REVERT` opcodes.